### PR TITLE
Move to nanobind

### PR DESCRIPTION
**Pull request by jmalkin**
Not super excited with how import_arrays() is added currently -- it seems like it's supposed to be part of a very short main() but we don't really have that in this case. Expecting I'll need to make additional changes here.

Vector of KLL might have some very slight performance regressions. I tried changing the get_index code back to std::vector and that didn't actually help. My current guess is that the more generic nanobind matrix data model has an additional cost. Whether we think there's enough other benefit to justify this will be the key deciding factor.

### CODE REVIEW

**Comment by jmalkin** [comment]
Created a simple, non-exhaustive test script. It uses python iteration to update since we expect that vectorized updates handled in c++ will be much closer in performance/.

Script:
```Python
import datasketches as ds
import numpy as np
import timeit

num_repeats = 100
n = 2 ** 20 # num updates

# default k in all cases since the test is of the interface

# cpc
def cpc_test():
  cpc = ds.cpc_sketch()
  for i in range(0, n):
    cpc.update(i)
  cpc.get_estimate()

  b = cpc.serialize()
  ds.cpc_sketch.deserialize(b)

# theta
def theta_test():
  sk = ds.update_theta_sketch()
  for i in range(0, n):
    sk.update(i)
  sk.get_estimate()

  b = sk.compact().serialize()
  ds.compact_theta_sketch.deserialize(b)

# tuple
def tuple_test():
  sk = ds.update_tuple_sketch(ds.AccumulatorPolicy())
  for i in range(0, n):
    sk.update(i, i * i)
  sk.get_estimate()

  b = sk.compact().serialize(ds.PyLongsSerDe())
  ds.compact_tuple_sketch.deserialize(b, ds.PyLongsSerDe())

# kll
def kll_test():
  kll = ds.kll_ints_sketch()
  for i in range(0, n):
    kll.update(i)
  kll.get_quantile(0.5)
  kll.get_rank(int(n / 2))

  b = kll.serialize()
  ds.kll_items_sketch.deserialize(b, ds.PyIntsSerDe())


# ebpps
def ebpps_test():
  ebpps = ds.ebpps_sketch(100)
  for i in range(0, n):
    ebpps.update(i)
  result = list(ebpps)

  b = ebpps.serialize(ds.PyIntsSerDe())
  ds.ebpps_sketch.deserialize(b, ds.PyIntsSerDe())


# density
def density_test():
  dim = 3
  sk = ds.density_sketch(k=50, dim=dim, kernel=ds.GaussianKernel())
  for i in range(0, int(n / 256)):
    sk.update(np.random.randn(dim))

  b = sk.serialize()
  ds.density_sketch.deserialize(b, ds.GaussianKernel())



elapsed = timeit.timeit(lambda: cpc_test(), number = num_repeats)
print(f'CPC: {elapsed:.4f}')

elapsed = timeit.timeit(lambda: theta_test(), number = num_repeats)
print(f'Theta: {elapsed:.4f}')

elapsed = timeit.timeit(lambda: tuple_test(), number = num_repeats)
print(f'Tuple: {elapsed:.4f}')

elapsed = timeit.timeit(lambda: kll_test(), number = num_repeats)
print(f'KLL: {elapsed:.4f}')

elapsed = timeit.timeit(lambda: ebpps_test(), number = num_repeats)
print(f'EBPPS: {elapsed:.4f}')

elapsed = timeit.timeit(lambda: density_test(), number = num_repeats)
print(f'Density: {elapsed:.4f}')
```

**Comment by jmalkin** [comment]
Pybind11 results:
```
CPC: 13.3239
Theta: 13.5307
Tuple: 21.6532
KLL: 15.2251
EBPPS: 21.6534
Density: 67.9749
```

Nanobind results:
```
CPC: 4.0155
Theta: 3.5051
Tuple: 7.3076
KLL: 4.5196
EBPPS: 15.3917
Density: 64.4 [rerun in isolation after changes]
```

File: CMakeLists.txt [hunk 1]
```diff
@@ -115,7 +116,7 @@ cmake_policy(SET CMP0097 NEW)
 include(ExternalProject)
 ExternalProject_Add(datasketches
   GIT_REPOSITORY https://github.com/apache/datasketches-cpp.git
-  GIT_TAG 5.0.0
+  GIT_TAG master
```
Comment (AlexanderSaydakov) [inline CMakeLists.txt:119]: why?

**Review by AlexanderSaydakov — APPROVED** [review]

File: CMakeLists.txt (hunk 1 above)
Comment (jmalkin) [reply CMakeLists.txt:119]: Oh, good catch. KLL needed the fix in 5.0.1 so need to flip it back to that.

Comment (jmalkin) [reply CMakeLists.txt]: Fixed

**Comment by c-dickens** [comment]
Just to make sure I understand this correctly: the timings reported above significantly better than those initially reported for nanobind? If we thought the performance was mixed before and this was a blocker to switching, these results refute that and suggest that performance is improved using nanobind, so we should switch.

**Comment by jmalkin** [comment]
@c-dickens I addressed the 50% degradation in density, and after trying a bunch of stuff in vector_of_kll it's, well, still a small drop in performance. But certainly for our classic sketches it's a win.

As noted, this is doing all the iteration in python, which we know is slow, but it seems to indicate that nanobind's python<->C++ interface really does provide less overhead for non-matrix/vector types.

**Comment by wjakob** [comment]
@jmalkin : I stumbled across this randomly. Can you see if the `NOMINSIZE` flag (see the bottom of https://nanobind.readthedocs.io/en/latest/building.html for an explanation) fixes your perf. regression?

**Comment by jmalkin** [comment]
@wjakob That did provide a benefit for most sketches in my limited testing (namely, rerun the above script).

I didn't try undoing the changes to use a numpy array constructor directly for one class (density_sketch) but part of the issue there is that the underlying C++ class wants to store vectors in std::vector as opposed to a more python-friendly type. We'll look at making that type a template parameter in the future.

Now I just need to revisit the object ownership details since I'm running into exactly the circular C++/Python dependency link described in the docs -- which isn't reported when running our unit tests!

**Pull request merged**